pkg/reconciler/v1alpha1/autoscaling/hpa/hpa.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hpa
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/knative/pkg/controller"
+	"github.com/knative/pkg/logging"
+	"github.com/knative/serving/pkg/apis/autoscaling"
+	pav1alpha1 "github.com/knative/serving/pkg/apis/autoscaling/v1alpha1"
+	informers "github.com/knative/serving/pkg/client/informers/externalversions/autoscaling/v1alpha1"
+	listers "github.com/knative/serving/pkg/client/listers/autoscaling/v1alpha1"
+	"github.com/knative/serving/pkg/reconciler"
+	"github.com/knative/serving/pkg/reconciler/v1alpha1/autoscaling/hpa/resources"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	autoscalingv1informers "k8s.io/client-go/informers/autoscaling/v1"
+	autoscalingv1listers "k8s.io/client-go/listers/autoscaling/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	controllerAgentName = "hpa-class-podautoscaler-controller"
+)
+
+type Reconciler struct {
+	*reconciler.Base
+
+	paLister  listers.PodAutoscalerLister
+	hpaLister autoscalingv1listers.HorizontalPodAutoscalerLister
+}
+
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+func NewController(
+	opts *reconciler.Options,
+	paInformer informers.PodAutoscalerInformer,
+	hpaInformer autoscalingv1informers.HorizontalPodAutoscalerInformer,
+) *controller.Impl {
+	c := &Reconciler{
+		Base:      reconciler.NewBase(*opts, controllerAgentName),
+		paLister:  paInformer.Lister(),
+		hpaLister: hpaInformer.Lister(),
+	}
+	impl := controller.NewImpl(c, c.Logger, "HPA-Class Autoscaling", reconciler.MustNewStatsReporter("HPA-Class Autoscaling", c.Logger))
+	c.Logger.Info("Setting up hpa-class event handlers")
+	paInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    impl.Enqueue,
+		UpdateFunc: controller.PassNew(impl.Enqueue),
+		DeleteFunc: impl.Enqueue,
+	})
+	return impl
+}
+
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key %s: %v", key, err))
+		return nil
+	}
+	logger := logging.FromContext(ctx)
+	logger.Debug("Reconcile hpa-clas PodAutoscaler")
+
+	original, err := c.paLister.PodAutoscalers(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		logger.Debug("PA no longer exists")
+		return c.deleteHpa(ctx, key)
```

Comment:
Here and below, I think we should set up the invariants such that a PodAutoscaler never changes classes.

This line is problematic because it could create a requeue spin loop when deletions fail (e.g. on NotFound).

Reply:
`c.deleteHpa` handles NotFound (search for "This is fine").  So think this is okay.  Am I missing something else?